packages/contractkit/src/identity/offchain/schemas/utils.ts

```diff
@@ -0,0 +1,348 @@
+    computedDataPath
+  )
+
+  if (!encryptedPayload.ok) {
+    return Err(new OffchainError(encryptedPayload.error))
+  }
+
+  const payload = await wallet.decrypt(readerPublicKeyAddress, encryptedPayload.result)
+  return Ok(payload)
+}
+
+/**
+ * Reads and decrypts a payload that has been encrypted to your data encryption key. Will
+ * resolving the symmetric key used to encrypt the payload.
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param senderAddress the address that encrypted this key to you
+ */
+export const readEncrypted = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  senderAddress: string
+): Promise<Result<Buffer, SchemaErrors>> => {
+  const encryptedPayloadPath = `${dataPath}.enc`
+  const [payload, key] = await Promise.all([
+    wrapper.readDataFromAsResult(
+      senderAddress,
+      (buf) => buildEIP712TypedData(wrapper, encryptedPayloadPath, buf),
+      encryptedPayloadPath
+    ),
+    readSymmetricKey(wrapper, dataPath, senderAddress),
+  ])
+
+  if (!payload.ok) {
+    return Err(new OffchainError(payload.error))
+  }
+  if (!key.ok) {
+    return Err(key.error)
+  }
+
+  if (key.result.length !== KEY_LENGTH) {
+    return Err(new InvalidKey())
+  }
+
+  return Ok(
+    AES128Decrypt(key.result, payload.result.slice(0, IV_LENGTH), payload.result.slice(IV_LENGTH))
+  )
+}
+
+export const deserialize = <DataType>(
+  type: t.Type<DataType>,
+  buf: Buffer
+): Result<DataType, SchemaErrors> => {
+  const dataAsJson = parseJsonAsResult(buf.toString())
+  if (!dataAsJson.ok) {
+    return Err(new InvalidDataError())
+  }
+
+  const parsedDataAsType = type.decode(dataAsJson.result)
+  if (isLeft(parsedDataAsType)) {
+    return Err(new InvalidDataError())
+  }
+
+  return Ok(parsedDataAsType.right)
+}
+
+export const buildEIP712TypedData = async <DataType>(
+  wrapper: OffchainDataWrapper,
+  path: string,
+  data: DataType | Buffer,
+  type?: t.Type<DataType>
+): Promise<EIP712TypedData> => {
+  const chainId = await wrapper.kit.web3.eth.getChainId()
+  const EIP712Domain = [
+    { name: 'name', type: 'string' },
+    { name: 'version', type: 'string' },
+    { name: 'chainId', type: 'uint256' },
+  ]
+
+  let types = {}
+  let message = {}
+  if (Buffer.isBuffer(data)) {
+    types = {
+      ClaimWithPath: [
+        { name: 'path', type: 'string' },
+        { name: 'hash', type: 'string' },
+      ],
+    }
+    message = {
+      hash: keccak256(data).toString('hex'),
+    }
+  } else {
+    const Claim = buildEIP712Schema(type!)
+    types = {
+      Claim,
+      ClaimWithPath: [
+        { name: 'path', type: 'string' },
+        { name: 'payload', type: 'Claim' },
+      ],
+    }
+    message = {
+      payload: (data as unknown) as EIP712Object,
+    }
+  }
+
+  return {
+    types: {
+      EIP712Domain,
+      ...types,
+    },
+    domain: {
+      name: 'CIP8 Claim',
```

Comment:
Is the term "claim" used here as a carry-over from the metadata system? If so, we may want to change it to "message" or "published data"

Reply:
I'm torn, I agree it doesn't make so much sense anymore all the time. But we _could_ be publicly/privately claiming a name/picture, in this case it still makes sense. Generally though (like publishing some private data associated with a transaction) it doesn't make sense

Reply:
I would consider a "claim" to be a conceptual subset of "published data" (since a claim must be published, but published data need not be a claim). So I would advocate for calling this something more generic to highlight that we intend for this to be a general purpose storage framework for dApps.